### PR TITLE
Fail if quorumSize is 1 and there are >2 leaders

### PR DESCRIPTION
**Pull request by jnewman2501**
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2184)

### CODE REVIEW

File: atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java [hunk 1]
```diff
@@ -79,6 +79,8 @@ public long leaderPingResponseWaitMs() {
 
     @Value.Check
     protected final void check() {
+        Preconditions.checkState(quorumSize() == 1 && leaders().size() > 2,
```
Comment (tpetracca) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: The only time quorom should be 1 is if there is exactly 1 leader, so should be `leaders().size() > 1`

**Comment by ashrayjain** [comment]
@jeremyk-91 @nziebart thoughts on removing the quromSize field entirely and just inferring it from the leaders list?

File: atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java (hunk 1 above)
Comment (tpetracca) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: Wait, doesn't check state ensure a thing is true? Shouldn't it actually be
```
if (quorumSize() == 1) {
    Preconditions.checkState(leaders().size() == 1,
```

Comment (tpetracca) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: which I feel like at the end of the day the check should just be
```
Preconditions.checkState(quorumSize() > leaders().size() / 2,
```

**Comment by jnewman2501** [comment]
+1 on inferring

File: atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java (hunk 1 above)
Comment (jnewman2501) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: Good point

**Comment by jnewman2501** [comment]
Waiting on ack on removing quorum size entirely before updating this PR

**Comment by jeremyk-91** [comment]
@ashrayjain Main question is if anyone would, for some reason, want to run more than a simple majority for their quorums. Don't see an immediate reason for this, but maybe large internal product uses it?

Also, see internal document on "TimeLock Server - Load Distribution" proposal 2 for where I'm thinking of taking this longer term.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java (hunk 1 above)
Comment (nziebart) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: This check is already covered below (`Preconditions.checkState(quorumSize() > leaders().size() / 2,`)

**Comment by nziebart** [comment]
I think if you want to expand from 3 to 5 nodes, you need to do some fiddling with the quorum size. Something along the lines of, first up quorum size to 3, then add 4th node, wait for paxos logs to propagate, then add 5th node.

**Comment by ashrayjain** [comment]
Not sure how important that is given we are moving to Timelock though

**Comment by ashrayjain** [comment]
Also i don't think large internal product uses embedded time and lock services

**Comment by ashrayjain** [comment]
Maybe @jboreiko or @clockfort have thoughts here?

**Comment by nziebart** [comment]
ah, forgot this was embedded leader. yeah, we might just be able to infer it

File: atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java (hunk 1 above)
Comment (jboreiko) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/config/LeaderConfig.java]: this ^

**Comment by jnewman2501** [comment]
Can someone take over this PR and make the desired changes?

**Comment by nziebart** [comment]
@jnewman2501 what was the original goal for this PR? I feel like the comments have gotten sidetracked

**Comment by jnewman2501** [comment]
@nziebart - original use case is that we want to support a sane default configuration out of the box that works for the non-HA case, but when people transition from non-HA -> HA they don't have a system that is set up to be unreliable (i.e. a quorum size of 1 with 3 nodes is WORSE than a quorum size of 1 with 1 node).

The default configuration we'd like to ship with looks like:

```
  atlas-leader:
    leaders: "{{discovered.all-stemma-api-uris}}"
    localServer: "{{self_discovered.stemma-api.uri}}"
    quorumSize: 1
    sslConfiguration:
      trustStorePath: "{{ssl.truststore_path}}"
      keyStorePath: "{{ssl.keystore_path}}"
      keyStorePassword: "{{ssl.keystore_password}}"
```

**Comment by tpetracca** [comment]
@jnewman2501 why can't you do that today? Like @nziebart mentions there is already a
```
Preconditions.checkState(quorumSize() > leaders().size() / 2,
                  "The quorumSize '%s' must be over half the amount of leader entries %s.", quorumSize(), leaders());
```
which should cover exactly your use case. Hard-coded quorum of 1 with 1 actual service works. Then we add 3 nodes but quorum is still 1. `1 > 1` is false and thus the checkState throw IllegalStateException.

**Comment by jnewman2501** [comment]
You are correct @tpetracca, closing out.

**Pull request closed without merging**